From 29dce8a5a7878cd0863373255889d3dff79906c9 Mon Sep 17 00:00:00 2001
From: Luca Bacci <luca.bacci982@gmail.com>
Subject: [PATCH] gconstructor: Reference _tls_used also on GCC, CLang

Backport the upstream GLib fix for missing Windows TLS callbacks with recent
mingw-w64 CRTs. Retaining the callback pointer alone does not retain the PE
TLS directory. The missing directory skips process/thread lifecycle callbacks.

Upstream: https://github.com/GNOME/glib/commit/29dce8a5a7878cd0863373255889d3dff79906c9
Fixes: https://gitlab.gnome.org/GNOME/glib/-/issues/4020
---
 glib/gconstructorprivate.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/glib/gconstructorprivate.h b/glib/gconstructorprivate.h
--- a/glib/gconstructorprivate.h
+++ b/glib/gconstructorprivate.h
@@ -56,6 +56,11 @@ __pragma (comment (linker, "/INCLUDE:" G_MSVC_SYMBOL_PREFIX "_ptr_" #func))
 
 #define G_HAS_TLS_CALLBACKS 1
 
+extern const IMAGE_TLS_DIRECTORY _tls_used;
+
+__attribute__ ((used, selectany))
+const IMAGE_TLS_DIRECTORY * const g_priv_tls_used_ = &_tls_used;
+
 #define G_DEFINE_TLS_CALLBACK(func) \
 static void NTAPI func (PVOID, DWORD, PVOID);          \
                                                        \
